main.go

```diff
@@ -36,12 +51,51 @@ type Firehose struct {
 	processSettings ProcessSettings
 	storageBackend  StorageBackend
 	hedwigConsumer  *hedwig.QueueConsumer
+	messageBuffer []*hedwig.Message
+	flushLock sync.Mutex
+	flushCh chan struct{}
+}
+
+func (fp *Firehose) WriteMessage(wg *sync.WaitGroup, message *hedwig.Message) error {
+	defer wg.Done()
+	// TODO: serialization for hedwig.Message to []byte
+	var data []byte
+	// TODO: how to get store appenv? should file name end with time?
+	currentTime := time.Now()
+	uploadLocation := fmt.Sprintf("%s/%s/%s/%s", "store-appenv?", message.Type, currentTime.Format("2006/1/2"), "filetime?")
+	return fp.storageBackend.UploadFile(context.Background(), data, fp.processSettings.StagingBucket, uploadLocation)
+}
+
+func (fp *Firehose) flushCron() {
+	<-time.After(time.Duration(fp.processSettings.FlushAfter) * time.Second)
+	fp.flushLock.Lock()
+	defer fp.flushLock.Unlock()
+	var wg sync.WaitGroup
+	for _, msg := range fp.messageBuffer {
+		wg.Add(1)
+		fp.WriteMessage(&wg, msg)
```

Comment:
does there have to be some way that if one message fails to write just that `handleMessage` should nack?  Or does the whole flush fail if one doesn't write correctly?

Reply:
Could fail the whole flush operation since that's simpler. Nacking a specific message would mean you want to get a channel per-message so you can send back feedback on it. Which, could work as well I suppose. Try it out?

Reply:
I'll keep it simple for now and add that in a later diff and task #11